data/coinductive.lean

```diff
@@ -0,0 +1,712 @@
+      rw ← Hk,
+      apply head_succ' _ _ g Hg, },
+    let g' := λ n, children' (g $ succ n) (cast (by rw this) i),
+    apply @ps_ih _ g',
+    { simp [g',all_agree], clear_except Hg,
+      intro,
+      generalize Hj : cast _ i = j,
+      generalize Hk : cast _ i = k,
+      have Hjk : j == k, cc, clear Hj Hk,
+      specialize Hg (succ n),
+      cases (g (succ n)), cases (g (succ (succ n))),
+      simp [children'], simp [agree] at Hg,
+      apply Hg.2 _ _ Hjk, },
+    intro k,
+    have Hsub_k := Hsub (k),
+    cases Hk_succ : g (k + length (sigma.mk y i :: ps_tl)) with _ y₂ ch₂,
+    simp [Hk_succ] at Hsub_k,
+    cases Hsub_k with _ Hsub_k, subst y,
+    simp [g'],
+    refine cast _ Hsub_k,
+    congr,
+    change g (succ $ k + length ps_tl) = _ at Hk_succ,
+    generalize Hj : cast _ i = j,
+    generalize Hk : cast _ i = k,
+    have Hjk : j == k, cc, clear Hj Hk,
+    revert k Hk_succ,
+    clear_except , generalize : (g (succ (k + length ps_tl))) = z,
+    intros, subst z, simp [children'], cases Hjk, refl }
+end
+
+@[simp]
+lemma roption.get_return {α} (x : α) (H)
+: roption.get (return x) H = x :=
+rfl
+
+lemma ext_aux {n : ℕ} (x y : cofix_a β (succ n)) (z : cofix_a β n)
+  (hx : agree z x)
+  (hy : agree z y)
+  (hrec : ∀ (ps : path' β),
+             (select' x ps).dom →
+             (select' y ps).dom →
+             n = ps.length →
+            (select' x ps) = (select' y ps))
+: x = y :=
+begin
+  induction n with n,
+  { cases x, cases y, cases z,
+    suffices : x_a = y_a,
+    { congr, assumption, subst y_a, simp,
+      funext i, cases x_a_1 i, cases y_a_1 i, refl },
+    clear hx hy,
+    specialize hrec [] trivial trivial rfl,
+    simp [select'] at hrec, injection hrec,
+    replace h_2 := congr_fun (eq_of_heq h_2) trivial,
+    exact h_2,  },
+  { cases x, cases y, cases z,
+    have : y_a = z_a,
+    { simp [agree] at hx hy, cc, },
+    have : x_a = y_a,
+    { simp [agree] at hx hy, cc, },
+    subst x_a, subst z_a, congr,
+    funext i, simp [agree] at hx hy,
+    apply n_ih _ _ (z_a_1 i),
+    { apply hx _ _ rfl },
+    { apply hy _ _ rfl },
+    { intros,
+      have : succ n = 1 + length ps,
+      { simp [*,one_add], },
+      have Hselect : ∀ (x_a_1 : β y_a → cofix_a β (succ n)),
+        (select' (cofix_a.intro y_a x_a_1) (⟨y_a, i⟩ :: ps)) = (select' (x_a_1 i) ps),
+      { rw this, simp [select_cons'], },
+      specialize hrec (⟨ y_a, i⟩ :: ps) _ _ (♯ this)
+      ; try { simp [Hselect,*], },
+      { simp [select',assert_if_pos] at hrec, exact hrec }, }, }
+end
+
+end coinduction.approx
+open coinduction.approx
+
+structure cofix_intl  {α : Type u} (β : α → Type v) : Type (max u v) :=
+  (approx : ∀ n, cofix_a β n)
+  (consistent : all_agree approx)
+
+def cofix (β : α → Type*) := cofix_intl β
+
+namespace cofix
+
+variables {X : Type*}
+variables (f : X → Σ y, β y → X)
+variables {β}
+protected def corec (i : X) : cofix β :=
+{ approx := s_corec f i
+, consistent := P_corec _ _ }
```

Comment:
I won't complain about coding style. https://github.com/leanprover/mathlib/blob/master/docs/style.md

Reply:
I think the naming is good but the spacing probably needs work. The proofs I think need a lot of work. A lot are far too long. I'll be addressing both as I whip the package into shape